### PR TITLE
gateway: add opentelemetry

### PR DESCRIPTION
**Pull request by ghost**
This PR adds a basic opentelemetry implementation on OSRD stack.

- [X] Basic opentelemetry in the gateway
  - [X] Support for Jaeger
  - [X] Support for Datadog
  - [X] Tracing
- [X] Add a jaeger "all-in-one" in the local docker compose

image

### CODE REVIEW

File: docker/gateway.dev.simple.toml [hunk 1]
```diff
@@ -6,12 +6,18 @@ secret_key = "NOT+A+SECRET++NOT+A+SECRET++NOT+A+SECRET++NOT+A+SECRET++NOT+A+SECR
 
 listen_addr = "0.0.0.0"
 
+[telemetry]
+type = "Jaeger"
+endpoint = "osrd-jaeger:6831"
+
 [[targets]]
+tracing_name = "editoast"
```
Comment (multun) [inline docker/gateway.dev.simple.toml:14]: how about just `name`? It's not like we couldn't use it for something else too

File: gateway/actix_proxy/Cargo.toml [hunk 2]
```diff
@@ -29,3 +29,7 @@ actix.workspace = true
 actix-web-actors.workspace = true
 bytestring.workspace = true
 
+# tracing / metrics
+opentelemetry.workspace = true
+actix-web-opentelemetry.workspace = true
+opentelemetry_sdk.workspace = true
```
Comment (multun) [inline gateway/actix_proxy/Cargo.toml]: missing newline

**Review by multun — APPROVED** [review]

File: gateway/src/main.rs [hunk 3]
```diff
@@ -32,6 +36,54 @@ async fn main() -> std::io::Result<()> {
         exit(1);
     });
 
+    let provider: Option<TracerProvider> = match config.clone().telemetry {
```
Comment (multun) [inline gateway/src/main.rs]: can you move it into the file where we do config parsing? `main()` is cluttered enough already

**Review by multun — CHANGES_REQUESTED** [review]

**Review by woshilapin — APPROVED** [review]
It's awesome that you tackled this 🚀 

2 general remarks at this point:
- why do we need some specific Jaeger/Datadog crates? Isn't the goal of OpenTelemetry to have a standard and not be vendor dependent? I used to connect to my OpenTelemetry collectors with only [opentelemetry_otlp](http://docs.rs/opentelemetry_otlp).
- I see that all the spans and attributes are done manually with the rather verbose API of `opentelemetry` crates. On top of that, that means that the `Context` has to be forwarded into every function call, which doesn't make it easy. Maybe we could leverage [`tracing`](https://docs.rs/tracing/latest/tracing/) and its ecosystem to simplify all of that. With `tracing`, the code would look a bit like the following.

```rust
#[tracing::instrument(
    skip_all,
    fields(
        proxy.upstream_uri = self.proxy.rebase_uri(Some("ws"), unprocessed_path, query).to_string(),
        proxy.type = "ws"
    )
)]
async fn handle_websocket(
        &self,
        req: &HttpRequest,
        stream: web::Payload,
        unprocessed_path: &str,
        query: &str,
    ) -> Result<HttpResponse, actix_web::Error> {
        // keep the original body without modification
}
```

This means no more `Context` (I believe `tracing` use the thread local to keep all the needed information). And all the information related to instrumentation of the code is not intertwined anymore with the body of the function. By default, a new span is created for each `#[tracing::instrument]` with the function's name used for the span's name (can be customized if needed).

I might be able to help for both remarks, so reach out if you need to. I believe the second (or even both?) remarks could be addressed later (which is why I'm fine to merge as-is since it already brings a very nice improvement on current situation).

**Comment by ghost** [comment]
> * why do we need some specific Jaeger/Datadog crates? Isn't the goal of OpenTelemetry to have a standard and not be vendor dependent? I used to connect to my OpenTelemetry collectors with only [opentelemetry_otlp](http://docs.rs/opentelemetry_otlp).

Removed the jaeger backend, implemented the otlp one. Kept the Datadog backend, since adding using otlp with datadog requires adding a translation service.

> * I see that all the spans and attributes are done manually with the rather verbose API of `opentelemetry` crates. On top of that, that means that the `Context` has to be forwarded into every function call, which doesn't make it easy. Maybe we could leverage [`tracing`](https://docs.rs/tracing/latest/tracing/) and its ecosystem to simplify all of that. With `tracing`, the code would look a bit like the following.

Tracing relies on thread-local variables to attribute spans in an async context, which require careful context switching on each future poll. It can either be done manually or via a macro but either way failure to add context switching instrumentation causes span misattribution.

The current method is less likely to be misused and does not involve any macro.

**Review by multun — APPROVED** [review]

**Comment by woshilapin** [comment]
> Removed the jaeger backend, implemented the otlp one. Kept the Datadog backend, since adding using otlp with datadog requires adding a translation service.

The advised way to do that is actually to use the Opentelemetry Collector with one of the contributed exporter for Datadog (see [here](https://docs.datadoghq.com/opentelemetry/otel_collector_datadog_exporter/)). I see one some pros and one con:
- **pros**
  * as an open-source project, we only depend on a open protocol and do not include code for a specific vendor
  * Opentelemetry Collector will play the role of the Intermediate Representation: imagine later we also start to collect CPU and memory from the pods, or Postgresql statistics. In each case, we're going to need to figure out how Datadog will work in each case. If later we end up moving from Datadog to some other solution, all of these will have to be done again. With Opentelemetry Collector, only the configuration of the Collector will have to change and redirect to another vendor, the applications and services don't have to evolve a bit.
- **cons**
  - deploy and maintain one more web-service

> Tracing relies on thread-local variables to attribute spans in an async context, which require careful context switching on each future poll. It can either be done manually or via a macro but either way failure to add context switching instrumentation causes span misattribution.
> 
> The current method is less likely to be misused and does not involve any macro.

Got it! Let's not overdo it, for now!

**Comment by multun** [comment]
EDIT: breaking news: the datadog agent is needed anyway, so keeping the datadog exporter over the OLTP one is pointless. @ElysaSrc is making the change

**Review by multun — APPROVED** [review]

**Pull request merged**